wsmux/session.go

```diff
@@ -0,0 +1,334 @@
+package wsmux
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+/*
+TODO: Add ping and pong handlers
+*/
+
+const (
+	defaultQueueSize            = 20
+	defaultStreamQueueSize      = 20
+	defaultKeepAliveInterval    = 30 * time.Second
+	defaultStreamAcceptDeadline = 30 * time.Second
+)
+
+// Session implements net.Listener. Allows creating and acception multiplexed streams over ws
+type Session struct {
+	mu       sync.Mutex
+	streams  map[uint32]*stream
+	streamCh chan *stream
+	writes   chan frame
+	conn     *websocket.Conn
+
+	keepAliveInterval    time.Duration
+	streamAcceptDeadline time.Duration
+
+	logger Logger
+
+	nextID uint32
+
+	closed chan struct{} // nil channel
+
+	closeConn bool
+
+	remoteCloseCallback func()
+}
+
+func newSession(conn *websocket.Conn, server bool, conf Config) *Session {
+	s := &Session{}
+	s.conn = conn
+	s.streams = make(map[uint32]*stream)
+	s.streamCh = make(chan *stream, defaultStreamQueueSize)
```

Comment:
I would suggest a timeout instead of a `defaultStreamQueueSize` see comments futher down...